src/main/java/com/playmonumenta/networkrelay/RabbitMQManager.java

```diff
@@ -154,23 +178,73 @@ protected RabbitMQManager(RabbitMQManagerAbstractionInterface abstraction, Logge
+					if (root.getAsJsonPrimitive("online").isBoolean() && root.getAsJsonPrimitive("online").getAsBoolean() == false) {
 						isDestShutdown = true;
 						mDestinationLastHeartbeat.remove(source);
 						mDestinationHeartbeatData.remove(source);
 						sendDestOfflineEvent(source);
 					}
 				}
+
 				if (!isDestShutdown) {
-					updateShardOnline(source, pluginDataFinal);
-				}
+					boolean heartbeatDataPresent = mDestinationLastHeartbeat.containsKey(source);
 
-				mAbstraction.sendMessageEvent(channel, source, data);
+					if (pluginDataFinal != null) {
+						/* This message contained heartbeat data - record it */
+						mDestinationLastHeartbeat.put(source, Instant.now());
+						mDestinationHeartbeatData.put(source, pluginDataFinal);
+					}
+
+					if (heartbeatDataPresent) {
+						/* This shard was already marked online - deliver normally */
+						mAbstraction.sendMessageEvent(channel, source, data);
+					} else {
+						/* This shard was not known to be online until this message */
+						if (pluginDataFinal == null) {
+							/*
+							 * Got a message from this shard but unfortunately it doesn't contain any plugin data
+							 * (i.e. it's not a heartbeat message)
+							 * This can happen randomly when other traffic is happening and this receiving shard just started
+							 * Can't send the online event yet - there's no heartbeat data which plugins might depend on while handling the online event
+							 * Also can't deliver the message to plugins, since they may be doing state tracking based on online status
+							 *
+							 * Instead, queue the packet for later delivery, once we do receive a heartbeat message containing plugin data
+							 */
+
+							/* Get existing queue or create and insert a new one */
+							ArrayDeque<QueuedMessage> queue = mDestinationQueuedMessages.computeIfAbsent(source, (unused) -> new ArrayDeque<>());
+							queue.addLast(new QueuedMessage(channel, data));
+
+							//TODO: change log level? Maybe warning if the queue is big or contains old entries?
+							mLogger.warning("Queued packet from " + source + " as this shard has not received heartbeat data yet. Current queue size is " + queue.size());
+						} else {
+							/*
+							 * Got a message from this shard and it has plugin data - great!
+							 * Have everything needed to send online event and deliver the message
+							 */
+
+							mLogger.fine("Shard " + source + " is online");
+							mAbstraction.sendDestOnlineEvent(source);
+
+							/* Deliver this current message */
+							mAbstraction.sendMessageEvent(channel, source, data);
+
+							/* Check if there were any queued messages from before heartbeat data was available and deliver them */
+							ArrayDeque<QueuedMessage> queue = mDestinationQueuedMessages.remove(source);
+							if (queue != null) {
+								for (QueuedMessage msg : queue) {
+									mLogger.fine("Delivering queued message from " + source + " now that it is marked as online");
```

Comment:
Even at log level `fine`, I'd suggest changing this to a message before and after the messages are sent instead - new shards, or shards that were offline for a long time, can display thousands of missed messages.